### ISSUE TITLE
preserve_fp32_precision flag is accepted as true for OPs for which it is not valid

### ISSUE THREAD

**Issue by amahmudTT** [tenstorrent/tt-metal#11756]
**Background:**
The `preserve_fp32_precision` flag was introduced to enable full fp32 precision for SFPU-only OPs, meaning in cases that the data is unpacked directly into Dest using `copy_tile()`. It does so by setting the `unpack_dst_format` for CBs to fp32 instead of tf32, so that the unpacker puts fp32 datums from the CB directly into Dest unchanged (skipping SrcA/B regs which are limited to 19bits).

**Issue:**
The `preserve_fp32_precision` flag is not valid for all OPs, but there is no warning/error produced when it is used incorrectly. For example reduce_tile utilizes the FPU, which works on the srcA and srcB registers, and setting `preserve_fp32_precision=True` will set the data type of srcA and srcB to fp32 instead of tf32 and lead to wrong results.

Also, the question was raised whether or not the flag is allowed in cases where a compute kernel may use one LLK for which the flag is supported but a second one for which it is not. One such example could be reloading fp32 partials into the Dest (flag supported) in order to perform further accumulation using matmul (flag not supported). In these cases, since the flag works by changing the `unpack_dst_format` for the buffer, if the buffer is only used for unpacking to Dest then the flag should be allowed. If the buffer is used to unpack to Dest but also unpack into SrcA/B, then the flag is not allowed.

**To Reproduce:**
This was detected in this bug https://github.com/tenstorrent/tt-metal/issues/10337 and this example can be used. 

**Expected behavior:**
We need a mechanism to disallow this flag to be set to true for OPs for which it is not valid, and better documentation surrounding this flag.

**Comment by hschoi4448** [comment]
> We need a mechanism to disallow this flag to be set to true for OPs for which it is not valid.

What does the above sentence mean among the two options below?
1. If reduce_tile exists in the compute kernel, that compute kernel does not support the `preserve_fp32` feature.
2. In the compute kernel, if both `add_tiles` and `reduce_tile` are present, `preserve_fp32` is disabled only for `reduce_tile`, which doesn't support it, while `add_tiles` remains unaffected.

**Comment by amahmudTT** [comment]
> > We need a mechanism to disallow this flag to be set to true for OPs for which it is not valid.
> 
> What does the above sentence mean among the two options below?
> 
> 1. If reduce_tile exists in the compute kernel, that compute kernel does not support the `preserve_fp32` feature.
> 2. In the compute kernel, if both `add_tiles` and `reduce_tile` are present, `preserve_fp32` is disabled only for `reduce_tile`, which doesn't support it, while `add_tiles` remains unaffected.

add_tiles also uses FPU and is not compatible with the flag, but if you had some other function for which preserve_fp32_precision is valid, that one should work fine, at least that is what we aim for.

**Comment by rdjogoTT** [comment]
@razorback3 Is there a reason this needs to be P0, since it is not blocking anyone. Now that we've clarified how to use the flag properly this isn't the highest priority. (I will still get to it next after solving #11962 but can we push this to P1 for now?)

**Comment by dongjin-na** [comment]
@rdjogoTT,
As we discussed in #10267, I found that there are scenarios where both cases are used in the same compute kernel in fp32 mode.

The issue is that a single `preserve_fp32_precision` flag cannot satisfy both cases. When the flag is set to true, the compute API produces incorrect results when using an fp32 CB as input. On the other hand, if the flag is false, only 16 bits are copied when unpacking to the destination register.

**Comment by razorback3** [comment]
Dear @rdjogoTT,
Currently, we are working on a training workload in a single Wormhole.
We succeeded in running the whole workload without errors or problems but we are facing a model convergence issue.
For that purpose, we are trying to use the fp32 datatype for some operations, not only the fp32 accumulation but also the fp32 datatype itself.

The problem is that some of our ops are using both SFPU and FPU in a single kernel.
In order to preserve whole fp32 data bits for computation that exploits SFPU, we should turn on the preserve_fp32_precison flag.
However, FPU computations require turning off the preserve_fp32_precison flag for correct output.
This is the collision point and the reason why we are sticking this issue as p0.

We hope you understand this situation and we appreciate you for the past issues that are resolved quickly.

**Comment by rdjogoTT** [comment]
Due to the need to support the ability to write compute kernels that mix OPs that can and can't use the flag (eg binary FPU and unary SFPU), we will need to rework the flag. The current plan to tackle this is to make the flag apply on a per-CB basis, meaning each CB can be controlled individually. CBs that serve as input to unsupported OPs will have the flag off, without preventing the other OPs from having fp32 precision. A plan is still being worked on for how to prevent flag misuse using asserts or similar.

**Comment by razorback3** [comment]
> Due to the need to support the ability to write compute kernels that mix OPs that can and can't use the flag (eg binary FPU and unary SFPU), we will need to rework the flag. The current plan to tackle this is to make the flag apply on a per-CB basis, meaning each CB can be controlled individually. CBs that serve as input to unsupported OPs will have the flag off, without preventing the other OPs from having fp32 precision. A plan is still being worked on for how to prevent flag misuse using asserts or similar.

Do you or your team have exact timeline for this?

**Comment by rdjogoTT** [comment]
@razorback3 
I have implemented the discussed solution on my branch rd/preserve_fp32_precision. I tested it with my own typecast op and made all necessary changes on the TT side. I also looked into where the flag was used in Moreh files and made appropriate changes to accommodate for the new vector flag. However, I need your help with the Moreh program factory files to set the correct CBs to have the flag enabled. Can someone from the Moreh side help with this?

For example in `ttnn/cpp/ttnn/operations/moreh/moreh_matmul/device/moreh_matmul_program_factory.cpp` I have initialized a vector `vector<PreserveFP32Target> preserve_fp32_precision(NUM_CIRCULAR_BUFFERS, PreserveFP32Target::Disabled);` which represents all CBs having the flag disabled. To enable the flag you can individually program which CBs you want it for by setting `preserve_fp32_precision[cb_id] = PreserveFP32Target::DEST`.

**Comment by razorback3** [comment]
> @razorback3 
> I have implemented the discussed solution on my branch rd/preserve_fp32_precision. I tested it with my own typecast op and made all necessary changes on the TT side. I also looked into where the flag was used in Moreh files and made appropriate changes to accommodate for the new vector flag. However, I need your help with the Moreh program factory files to set the correct CBs to have the flag enabled. Can someone from the Moreh side help with this?
> 
> For example in `ttnn/cpp/ttnn/operations/moreh/moreh_matmul/device/moreh_matmul_program_factory.cpp` I have initialized a vector `vector<PreserveFP32Target> preserve_fp32_precision(NUM_CIRCULAR_BUFFERS, PreserveFP32Target::Disabled);` which represents all CBs having the flag disabled. To enable the flag you can individually program which CBs you want it for by setting `preserve_fp32_precision[cb_id] = PreserveFP32Target::DEST`.

Nice work!
@dongjin-na will check the functionality on the given branch and give feedback soon 😀.

**Comment by dongjin-na** [comment]
Dear @rdjogoTT,

I’ve reviewed the `moreh_matmul` case on your branch. 
The `test_moreh_matmul_fp32_dest_acc` passed successfully after setting `preserve_fp32_precision[tt::CB::c_intermed0] = PreserveFP32Target::DEST`. 
I’ll take a look at a few other cases on the Moreh side and will update you soon.

I do have one question, though. 
Is `PreserveFP32Target::SRCA_B` supported yet? When I set it, the JIT build failed with the following error:
```
Always | FATAL | preserve_fp32_precision is only available when unpack target is DEST register
```

Let me know if you need further adjustments. :)

**Comment by rdjogoTT** [comment]
@dongjin-na 
I forgot to mention that I added `PreserveFP32Target::SRCA_B` even though it is NOT supported. It isn't supported because if the CBs are being unpacked into SrcA/B, then fp32 precision is not possible (the best we can do if TF32). This is due to the limitations of the Src regs and FPU that we have discussed in earlier issues/threads.

I added `PreserveFP32Target::SRCA_B` intentionally so that programmers who are trying to use the flag can set the unpack target to SrcA/B or Dest and will hit this error if trying to use the flag incorrectly (with srca/b).

**Comment by dongjin-na** [comment]
@rdjogoTT, thanks for the quick response and detailed explanation.

I have one more question. 
In our softmax implementation, there is a case where both the SFPU and FPU use fp32 CBs.

```c++
    auto intermed_data_format = fp32_dest_acc_en ? tt::DataFormat::Float32 : data_format;
    // ......
    constexpr auto cb_exps = tt::CB::c_intermed0;
    constexpr auto cb_recipsumexps = tt::CB::c_intermed1;
    constexpr auto cb_add = tt::CB::c_intermed2;
    constexpr auto cb_max = tt::CB::c_intermed3;
    constexpr auto cb_tmp = tt::CB::c_intermed4;
```

https://github.com/tenstorrent/tt-metal/blob/0d20c99c3c8e86226e6ac70f09f5d4577cc0a25a/ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/moreh_softmax/kernels/moreh_softmax_w_large.cpp#L91-L140

Creating two CBs with different PreserveFP32Target values that share the same memory seems like a good approach to reducing precision loss as much as possible. The SFPU would use a CB with PreserveFP32Target::DEST enabled, while the FPU would use a CB with PreserveFP32Target::Disabled.

Do you think this approach is appropriate?

**Comment by dongjin-na** [comment]
> I'm not sure what you mean when you say that two CBs can "share the same memory". Does this mean that you will refer to a single memory space by two different CB names?

Yes, as in the example with [matmul](https://github.com/tenstorrent/tt-metal/blob/149c7584c0305c232f1a3cef9b8aa5c6bc5a131a/ttnn/cpp/ttnn/operations/matmul/device/matmul_op_multi_core_reuse_optimized_program_factory.cpp#L302-L307), two CBs can share the same memory space. That’s what I was referring to.

> Also, can you clarify what you want to accomplish for Softmax? Do you want to use a single CB as an input to both a FPU and SFPU OP?

In the softmax example mentioned above, the already implemented kernel code does have a scenario where a single CB is used as input to both an FPU and an SFPU. 
To achieve this, based on your PR, we can modify the kernel code to use a CB with `PreserveFP32Target::DEST` enabled for SFPU operations and a CB with `PreserveFP32Target::Disabled` for FPU operations.

I’ve checked the `PreserveFP32Target` on Moreh's side. 
Could you please merge these changes into the `main` branch?

**Comment by rdjogoTT** [comment]
@dongjin-na Ok makes sense. Do you want to push the missing more changes to my branch so I can merge them together with my changes, or do you want me to just push my changes and you will later fix the moreh files?

**Comment by razorback3** [comment]
@rdjogoTT

**Comment by razorback3** [comment]
@rdjogoTT
Please push your changes to the main first. Thanks for your work 👍

**Comment by rdjogoTT** [comment]
@razorback3 @dongjin-na can I get a review for the PR: https://github.com/tenstorrent/tt-metal/pull/12794

**Comment by razorback3** [comment]
@rdjogoTT
Seems like all works as intended. Salute to the nice work!
I think you can close this issue after the PR is merged to the main branch.

**Comment by rdjogoTT** [comment]
@razorback3 @dongjin-na I've updated the naming of the flag based on feedback from others within TT. The flag still works the exact same way as the preserve_fp32_precision vector I created, it just had a different name. The new enum is:

```
enum class UnpackToDestMode : uint8_t
{
    UnpackToDestFp32,
    Default
};
```

`Default` is the new value that corresponds to regular functionality, and `UnpackToDestFp32` needs to be set for CBs that you want to use to unpack fp32 data to Dest in full precision.
